### PR TITLE
Add type hints to `tests/rest`

### PR DESCRIPTION
**Pull request by dklimpel**
### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] Pull request includes a [sign off](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#sign-off)
* [x] [Code style](https://matrix-org.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://matrix-org.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

Signed-off-by: Dirk Klimpel dirk@klimpel.org

### CODE REVIEW

File: tests/rest/media/v1/test_url_preview.py [hunk 1]
```diff
@@ -140,9 +145,9 @@ def resolveHostName(
                 resolutionReceiver.resolutionComplete()
                 return resolutionReceiver
 
-        self.reactor.nameResolver = Resolver()
+        self.reactor.nameResolver = Resolver()  # type: ignore[assignment]
```
Comment (clokep) [inline tests/rest/media/v1/test_url_preview.py:148]: Rather annoying that `MemoryReactor` doesn't implement [`IReactorPluggableResolver`](https://twistedmatrix.com/documents/current/api/twisted.internet.interfaces.IReactorPluggableResolver.html). 👍

**Review by clokep — APPROVED** [review]
🎉 Great!

File: changelog.d/12208.misc [hunk 2]
```diff
@@ -0,0 +1 @@
+Add type hints to `tests/rest`.
```
Comment (clokep) [inline changelog.d/12208.misc:1]: In #12207 I made this a bit more generic:

```suggestion
Add type hints to tests files.
```

File: mypy.ini [hunk 3]
```diff
@@ -90,7 +90,6 @@ exclude = (?x)
    |tests/push/test_push_rule_evaluator.py
    |tests/rest/client/test_transactions.py
    |tests/rest/media/v1/test_media_storage.py
```
Comment (clokep) [inline mypy.ini:92]: Can we remove this file too?

Comment (dklimpel) [reply mypy.ini:92]: Unfortunately no. There are many more complex errors.

Comment (clokep) [reply mypy.ini:92]: No problem! Just making sure it wasn't an oversight!

**Pull request merged**